pkg/controller/clusterprovision/clusterprovision_controller.go

```diff
@@ -0,0 +1,319 @@
+
+	hivev1 "github.com/openshift/hive/pkg/apis/hive/v1alpha1"
+	controllerutils "github.com/openshift/hive/pkg/controller/utils"
+	"github.com/openshift/hive/pkg/install"
+)
+
+const (
+	controllerName = "clusterProvision"
+
+	adminSSHKeySecretKey = "ssh-publickey"
+
+	ClusterDeploymentLabelKey = "hive.openshift.io/cluster-deployment"
+)
+
+// Add creates a new ClusterProvision Controller and adds it to the Manager with default RBAC. The Manager will set fields on the Controller
+// and Start it when the Manager is Started.
+func Add(mgr manager.Manager) error {
+	return add(mgr, newReconciler(mgr))
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager) reconcile.Reconciler {
+	return &ReconcileClusterProvision{Client: controllerutils.NewClientWithMetricsOrDie(mgr, controllerName), scheme: mgr.GetScheme()}
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("clusterprovision-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to ClusterProvision
+	err = c.Watch(&source.Kind{Type: &hivev1.ClusterProvision{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to ClusterDeployment
+	err = c.Watch(&source.Kind{Type: &hivev1.ClusterDeployment{}}, &handler.EnqueueRequestsFromMapFunc{
+		ToRequests: handler.ToRequestsFunc(
+			func(a handler.MapObject) []reconcile.Request {
+				return selectorDeploymentWatchHandler(a, mgr)
+			}),
+	})
+
+	// Watch for install jobs created for ClusterProvision
+	err = c.Watch(&source.Kind{Type: &batchv1.Job{}}, &handler.EnqueueRequestForOwner{
+		IsController: true,
+		OwnerType:    &hivev1.ClusterProvision{},
+	})
+
+	return nil
+}
+
+var _ reconcile.Reconciler = &ReconcileClusterProvision{}
+
+// ReconcileClusterProvision reconciles a ClusterProvision object
+type ReconcileClusterProvision struct {
+	client.Client
+	scheme *runtime.Scheme
+}
+
+// Reconcile reads that state of the cluster for a ClusterProvision object and makes changes based on the state read
+// and what is in the ClusterProvision.Spec
+// +kubebuilder:rbac:groups=hive.openshift.io,resources=ClusterProvisions,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=hive.openshift.io,resources=ClusterProvisions/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=hive.openshift.io,resources=ClusterDeployment,verbs=get;list;watch
+// +kubebuilder:rbac:groups=batch,resources=jobs,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=core,resources=configmaps;namespaces,verbs=get
+func (r *ReconcileClusterProvision) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	pLog := log.WithFields(log.Fields{
+		"name":       request.NamespacedName.String(),
+		"controller": controllerName,
+	})
```

Comment:
Should add reconcile loop metrics similar to the other controllers here. (clusterdeployment is a good example)